### PR TITLE
Implement XEP-0167: Jingle RTP Sessions Informational Messages

### PR DESCRIPTION
**Pull request by melvo**
PR check list:
- [x] Document your code
- [x] Add `\since QXmpp 1.X`, `QXMPP_EXPORT`
- [x] Fix doxygen warnings (see log when building with `-DBUILD_DOCUMENTATION=ON`)
- [x] Update `doc/doap.xml`
- [x] Add unit tests
- [x] Format the code: Run `clang-format -i src/<edited-file(s)> tests/<edited-file(s)>`

### CODE REVIEW

File: src/base/QXmppJingleIq.cpp [hunk 1]
```diff
@@ -839,37 +851,76 @@ void QXmppJingleIq::setResponder(const QString &responder)
...
 ///
-/// \param ringing
+/// \param sid
 
-void QXmppJingleIq::setRinging(bool ringing)
+void QXmppJingleIq::setSid(const QString &sid)
 {
-    d->ringing = ringing;
+    d->sid = sid;
 }
```
Comment (lnjX) [inline src/base/QXmppJingleIq.cpp:867]: please either use
```
///
/// text
///
```
or
```
/// text
```

File: src/base/QXmppJingleIq.cpp [hunk 2]
```diff
@@ -839,37 +851,76 @@ void QXmppJingleIq::setResponder(const QString &responder)
...
+/// Informational Messages.
+///
+/// \return the session's state
+///
+/// \since QXmpp 1.5
+///
+QXmppJingleIq::RtpSessionState QXmppJingleIq::rtpSessionState() const
+{
+    return d->rtpSessionState;
+}
```
Comment (lnjX) [inline src/base/QXmppJingleIq.cpp:880]: doesn't seem to be tested

File: src/base/QXmppJingleIq.cpp [hunk 3]
```diff
@@ -839,37 +851,76 @@ void QXmppJingleIq::setResponder(const QString &responder)
...
-/// \param sid
+/// Returns true if the call is ringing.
+///
+/// \deprecated This method is deprecated since QXmpp 1.5. Use\c QXmpp::rtpSessionState()
+/// instead.
+///
+bool QXmppJingleIq::ringing() const
+{
+    return d->rtpSessionState == QXmppJingleIq::Ringing;
+}
```
Comment (lnjX) [inline src/base/QXmppJingleIq.cpp:905]: please keep it at its previous position

File: src/base/QXmppJingleIq.cpp [hunk 4]
```diff
@@ -904,9 +955,16 @@ void QXmppJingleIq::parseElementFromChild(const QDomElement &element)
...
 
-    // ringing
-    QDomElement ringingElement = jingleElement.firstChildElement(QStringLiteral("ringing"));
-    d->ringing = (ringingElement.namespaceURI() == ns_jingle_rtp_info);
+    for (auto childElement = jingleElement.firstChildElement();
+         !childElement.isNull();
+         childElement = childElement.nextSiblingElement()) {
+        if (childElement.namespaceURI() == ns_jingle_rtp_info) {
+            if (const auto rtpSessionState = JINGLE_RTP_SESSION_STATES.indexOf(childElement.tagName()); rtpSessionState != -1) {
+                d->rtpSessionState = static_cast<QXmppJingleIq::RtpSessionState>(rtpSessionState);
```
Comment (lnjX) [inline src/base/QXmppJingleIq.cpp:963]: no static cast needed, it's okay to use the functional style cast here

File: src/base/QXmppJingleIq.cpp [hunk 5]
```diff
@@ -922,9 +980,8 @@ void QXmppJingleIq::toXmlElementFromChild(QXmlStreamWriter *writer) const
     }
     d->reason.toXml(writer);
 
-    // ringing
-    if (d->ringing) {
-        writer->writeStartElement(QStringLiteral("ringing"));
+    if (const auto rtpSessionState = d->rtpSessionState; rtpSessionState != QXmppJingleIq::None) {
+        writer->writeStartElement(JINGLE_RTP_SESSION_STATES.at(rtpSessionState));
         writer->writeDefaultNamespace(ns_jingle_rtp_info);
         writer->writeEndElement();
```
Comment (lnjX) [inline src/base/QXmppJingleIq.cpp:986]: why are you copying into `const auto rtpSessionState` here? you can even have a simpler if statement if you don't do it

File: src/base/QXmppJingleIq.h [hunk 6]
```diff
@@ -296,13 +321,17 @@ class QXMPP_EXPORT QXmppJingleIq : public QXmppIq
...
     QString sid() const;
     void setSid(const QString &sid);
 
+    RtpSessionState rtpSessionState() const;
+    void setRtpSessionState(RtpSessionState rtpSessionState);
+
+#if QXMPP_DEPRECATED_SINCE(1, 5)
+    bool ringing() const;
+    void setRinging(bool ringing);
+#endif
```
Comment (lnjX) [inline src/base/QXmppJingleIq.h:333]: please also mark them as deprecated

**Review by lnjX — CHANGES_REQUESTED** [review]

File: src/base/QXmppJingleIq.h (hunk 6 above)
Comment (melvo) [reply src/base/QXmppJingleIq.h]: I took that from `QXmpp::Encryption`. Do we have to add `QT_DEPRECATED_X()` to `QXmpp::Encryption` as well?

File: src/base/QXmppJingleIq.cpp (hunk 2 above)
Comment (melvo) [reply src/base/QXmppJingleIq.cpp:910]: Yes, thus the pull request is in Draft state and the corresponding checkbox unchecked.

File: src/base/QXmppJingleIq.h (hunk 6 above)
Comment (lnjX) [reply src/base/QXmppJingleIq.h]: `QT_DEPRECATED_X()` does not work on constants, but there's prob also something for that, so would be good yes.

File: src/base/QXmppJingleIq.cpp [hunk 7]
```diff
@@ -839,24 +851,37 @@ void QXmppJingleIq::setResponder(const QString &responder)
     d->responder = responder;
 }
 
+#if QXMPP_DEPRECATED_SINCE(1, 5)
```
Comment (lnjX) [inline src/base/QXmppJingleIq.cpp:854]: not in the cpp file please, the function needs to be compiled in for users needing it

File: src/base/QXmppJingleIq.cpp [hunk 8]
```diff
@@ -56,6 +57,16 @@ static const char *jingle_reasons[] = {
     "unsupported-transports",
 };
 
+static const QStringList JINGLE_RTP_SESSION_STATES = {
+    QStringLiteral(),
```
Comment (lnjX) [inline src/base/QXmppJingleIq.cpp:61]:
```suggestion
    QString(),
```

File: src/base/QXmppJingleIq.cpp [hunk 9]
```diff
@@ -998,10 +1071,37 @@ void QXmppJingleIq::toXmlElementFromChild(QXmlStreamWriter *writer) const
...
+            }
+
+            helperToXmlAddAttribute(writer, QStringLiteral("name"), rtpSessionStateMuting->name);
+        } else {
+            writer->writeStartElement(QStringLiteral("ringing"));
+        }
+
+        if (!std::holds_alternative<RtpSessionStateMuting>(*d->rtpSessionState)) {
+            writer->writeDefaultNamespace(ns_jingle_rtp_info);
+        }
```
Comment (lnjX) [inline src/base/QXmppJingleIq.cpp:1103]: this is a bit subtle, please just put the `writeDefaultNamespace()` in every case

File: src/base/QXmppJingleIq.h [hunk 10]
```diff
@@ -157,6 +159,79 @@ class QXMPP_EXPORT QXmppJingleIq : public QXmppIq
         TransportReplace
     };
 
+    ///
+    /// Party muting or unmuting itself
+    ///
+    /// \since QXmpp 1.5
+    ///
```
Comment (lnjX) [inline src/base/QXmppJingleIq.h:166]: documentation into the `.cpp` please

File: src/base/QXmppJingleIq.cpp [hunk 11]
```diff
@@ -875,24 +876,39 @@ void QXmppJingleIq::setResponder(const QString &responder)
     d->responder = responder;
 }
 
+///
 /// Returns true if the call is ringing.
-
+///
+/// \deprecated This method is deprecated since QXmpp 1.5. Use \c QXmpp::rtpSessionState()
```
Comment (lnjX) [inline src/base/QXmppJingleIq.cpp:882]: `QXmpp::rtpSessionState()` seems wrong

File: src/base/QXmppJingleIq.cpp [hunk 12]
```diff
@@ -875,24 +876,39 @@ void QXmppJingleIq::setResponder(const QString &responder)
...
+    return false;
 }
 
+///
 /// Set to true if the call is ringing.
 ///
 /// \param ringing
-
+///
+/// \deprecated This method is deprecated since QXmpp 1.5. Use \c QXmpp::setRtpSessionState()
```
Comment (lnjX) [inline src/base/QXmppJingleIq.cpp:899]: `QXmpp::setRtpSessionState()`

File: src/base/QXmppJingleIq.h [hunk 13]
```diff
@@ -157,6 +159,79 @@ class QXMPP_EXPORT QXmppJingleIq : public QXmppIq
...
+    enum RtpSessionStateMutingCreator {
```
Comment (lnjX) [inline src/base/QXmppJingleIq.h:167]:
```suggestion
    enum Creator {
```
I don't know whether this is actually useful, but what do you think about using this enum for general purposes and calling it just `Creator`?

Comment (melvo) [reply src/base/QXmppJingleIq.h]: I replaced it by a bool.

Comment (lnjX) [reply src/base/QXmppJingleIq.h]: please use an enum, that's more explicit

File: src/base/QXmppJingleIq.cpp [hunk 14]
```diff
@@ -1063,9 +1159,36 @@ void QXmppJingleIq::parseElementFromChild(const QDomElement &element)
...
+                    muting.isInitiatorCreator = true;
+                } else if (creator == QStringLiteral("responder")) {
+                    muting.isInitiatorCreator = false;
+                }
+
+                muting.name = childElement.attribute(QStringLiteral("name"));
+
+                d->rtpSessionState = muting;
+            } else if (elementTag == QStringLiteral("ringing")) {
+                d->rtpSessionState = QXmppJingleIq::RtpSessionStateRinging();
```
Comment (lnjX) [inline src/base/QXmppJingleIq.cpp:1188]:
```suggestion
                d->rtpSessionState = RtpSessionStateRinging();
```
?

File: src/base/QXmppJingleIq.h (hunk 13 above)
Comment (melvo) [reply src/base/QXmppJingleIq.h]: I only renamed it now.

**Review by lnjX — APPROVED** [review]

**Pull request merged**